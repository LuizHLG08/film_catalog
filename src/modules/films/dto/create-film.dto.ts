import { IsNumber, IsString, MaxLength } from "class-validator";
export class CreateFilmDto {

    @IsString()
    name: string;

    @IsNumber()
    year: number;

    @IsNumber()
    price: number;

    @IsString()
    @MaxLength(500)
    description: string;

}
