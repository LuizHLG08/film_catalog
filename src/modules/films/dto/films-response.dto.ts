import { ApiProperty } from "@nestjs/swagger";

export class FilmsResponseDto {
    @ApiProperty({
        description: 'Nome do filme',
        example: 'Filme',
    })
    id: string;
    @ApiProperty({
        description: 'Nome do filme',
        example: 'Filme',
    })
    name: string;
    @ApiProperty({
        description: 'Nome do filme',
        example: 'Filme',
    })
    year: number;
    @ApiProperty({
        description: 'Nome do filme',
        example: 'Filme',
    })
    price: number;
    @ApiProperty({
        description: 'Nome do filme',
        example: 'Filme',
    })
    description: string;
    @ApiProperty({
        description: 'Nome do filme',
        example: 'Filme',
    })
    createdAt: string;
    @ApiProperty({
        description: 'Nome do filme',
        example: 'Filme',
    })
    updatedAt: string;
    @ApiProperty({
        description: 'Nome do filme',
        example: 'Filme',
    })
    ownerId?: string;

}