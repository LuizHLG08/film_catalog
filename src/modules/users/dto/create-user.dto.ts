import { hashSync } from "bcrypt";
import { Transform } from "class-transformer";
import { IsEmail, IsString, Matches, MinLength } from "class-validator";

export class CreateUserDto {

    @IsString()
    name: string;

    @IsEmail()
    email: string;

    @IsString()
    @MinLength(4)
    @Matches(/((?=.*\d)|(?=.*\W+))(?![.\n])(?=.*[A-Z])(?=.*[a-z]).*$/, {
        message: 'password too weak',
    })
    @Transform(({value}: {value:string}) => hashSync(value, 10), {groups: ['transform']})
    password: string;
}
