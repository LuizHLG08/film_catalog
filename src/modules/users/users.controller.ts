import { Controller, Get, Post, Body, Patch, Param, Delete, Res, Req } from '@nestjs/common';
import { UsersService } from './users.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { Response } from 'express';
import { IsPublic } from 'src/auth/decorators/is-public.decorator';
import { AuthRequest } from 'src/auth/models/auth.request';

@Controller('users')
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @IsPublic()
  @Post()
  async create(@Body() createUserDto: CreateUserDto, @Res() res: Response) {
    return res.status(201).json(await this.usersService.create(createUserDto));
  }

  @Get()
  async findAll(@Res() res: Response) {
    return res.status(200).json(await this.usersService.findAll());
  }

  @Patch()
  async update(@Req() req : AuthRequest, @Body() updateUserDto: UpdateUserDto, @Res() res: Response) {
    return res.status(200).json(await this.usersService.update(req.user.id, updateUserDto));
  }

  @Delete()
  async remove(@Req() req : AuthRequest, @Res() res: Response) {
    await this.usersService.remove(req.user.id)
    return res.status(204);
  }
}
