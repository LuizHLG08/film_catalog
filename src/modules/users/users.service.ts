import { Injectable } from '@nestjs/common';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { InjectRepository } from '@nestjs/typeorm';
import { Users } from './entities/users.entity';
import { Repository } from 'typeorm';

@Injectable()
export class UsersService {

  constructor(@InjectRepository(Users) private readonly usersRepository: Repository<Users>) {}

  async create(createUserDto: CreateUserDto) : Promise<Users> {
    const user = await this.usersRepository.save(createUserDto)
    return {
      ...user,
      password: undefined
    };
  }

  async findAll(): Promise<Users[]> {
    const users = await this.usersRepository.find();
    const responseUsers = users.map(user => {
      return {
        ...user,
        password: undefined
      }
    })
    return responseUsers;
  }

  async findByEmail(email: string) {
    const user = await this.usersRepository.findOne({where: { email : email }});

    return user
  }

  async update(id: string, updateUserDto: UpdateUserDto) {
    const updateResult = await this.usersRepository.update({ id }, updateUserDto);

    if (updateResult.affected > 0) {
        const updatedUser = await this.usersRepository.findOne({where : {id : id}});
        return {
          ...updatedUser,
          password: undefined,
          updatedAt: new Date()
        };
    } else {
        throw new Error('User not found or not updated.');
    }
  }

  async remove(id: string) {
    return await this.usersRepository.delete(id);
  }
}
