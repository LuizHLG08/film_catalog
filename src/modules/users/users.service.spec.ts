import { Test, TestingModule } from '@nestjs/testing';
import { UsersService } from './users.service';
import { Users } from './entities/users.entity';
import { Repository } from 'typeorm';
import { getRepositoryToken } from '@nestjs/typeorm';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';

const mockUserDto: CreateUserDto = {
  name: 'Test User',
  email: 'test@example.com',
  password: 'password123',
};

const mockUser: Users = new Users({
  id: '1',
  name: 'Test User',
  email: 'test@example.com',
  password: 'password123',
});

describe('UsersService', () => {
  let usersService: UsersService;
  let userRepository: Repository<Users>;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        UsersService,
        {
          provide: getRepositoryToken(Users),
          useValue: {
            save: jest.fn().mockResolvedValue(mockUser),
            find: jest.fn().mockResolvedValue([mockUser]),
            findOne: jest.fn().mockResolvedValue(mockUser),
            update: jest.fn().mockResolvedValue(true),
            delete: jest.fn().mockResolvedValue(true),
          },
        },
      ],
    }).compile();

    usersService = module.get<UsersService>(UsersService);
    userRepository = module.get<Repository<Users>>(getRepositoryToken(Users));
  });

  it('should be defined', () => {
    expect(usersService).toBeDefined();
    expect(userRepository).toBeDefined();
  });

  describe('create', () => {
    it('should be able to create users correctly', async () => {
      const createdUser = await usersService.create(mockUserDto);

      expect(userRepository.save).toHaveBeenCalledWith(mockUserDto);
      expect(createdUser).toEqual({
        ...mockUser,
        password: undefined,
      });
    });
  });

  describe('findAll', () => {
    it('should return a users list successfully', async () => {
      const usersList = await usersService.findAll();

      expect(userRepository.find).toHaveBeenCalled();
      expect(usersList).toEqual([{
        ...mockUser,
        password: undefined,
      }]);
    });
  });

  describe('update', () => {
    it('should be able to update users correctly', async () => {
      const updateUserDto: UpdateUserDto = { name: 'Updated Name' };
  
      const mockUpdatedUser = { ...mockUser, ...updateUserDto };
      userRepository.update = jest.fn().mockResolvedValue({ affected: 1 });
      userRepository.findOne = jest.fn().mockResolvedValue(mockUpdatedUser);
  
      const updatedUser = await usersService.update('1', updateUserDto);
  
      expect(userRepository.update).toHaveBeenCalledWith({ id: '1' }, updateUserDto);
      expect(userRepository.findOne).toHaveBeenCalledWith({ where: { id: '1' } });
      expect(updatedUser).toEqual({
        ...mockUpdatedUser,
        password: undefined,
      });
    });
  
    it('should throw an error when user is not found or not updated', async () => {
      userRepository.update = jest.fn().mockResolvedValue({ affected: 0 });
  
      await expect(usersService.update('1', {} as UpdateUserDto)).rejects.toThrowError('User not found or not updated.');
    });
  });

  describe('remove', () => {
    it('should be able to delete users correctly', async () => {
      await usersService.remove('1');

      expect(userRepository.delete).toHaveBeenCalledWith('1');
    });
  });
});
