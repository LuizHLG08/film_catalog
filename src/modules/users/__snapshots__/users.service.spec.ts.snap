// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersService update should throw an error when user is not found or not updated: User not found or not updated. 1`] = `"User not found or not updated."`;
